lua/null-ls/builtins/formatting/nix_flake_fmt.lua

```diff
@@ -0,0 +1,151 @@
+    -- once the [Lazy trees PR] lands.
+    --
+    -- [Lazy trees PR]: https://github.com/NixOS/nix/pull/6530
+    local cp = vim.system({ "nix", "--extra-experimental-features", "nix-command", "config", "show", "system" }):wait()
+    if cp.code ~= 0 then
+        log:warn(string.format("unable to discover builtins.currentSystem from nix. stderr: %s", cp.stderr))
+        return nil
+    end
+    local nix_current_system = cp.stdout:match("^(.-)%s+") -- remove trailing newline
+
+    local eval_nix_formatter = [[
+      let
+        currentSystem = "]] .. nix_current_system .. [[";
+        # Various functions vendored from nixpkgs lib (to avoid adding a
+        # dependency on nixpkgs).
+        lib = rec {
+          getOutput = output: pkg:
+            if ! pkg ? outputSpecified || ! pkg.outputSpecified
+            then pkg.${output} or pkg.out or pkg
+            else pkg;
+          getBin = getOutput "bin";
+          # Simplified by removing various type assertions.
+          getExe' = x: y: "${getBin x}/bin/${y}";
+          # getExe is simplified to assume meta.mainProgram is specified.
+          getExe = x: getExe' x x.meta.mainProgram;
+        };
+      in
+      formatterBySystem:
+        if formatterBySystem ? ${currentSystem} then
+          let
+            formatter = formatterBySystem.${currentSystem};
+            drv = formatter.drvPath;
+            bin = lib.getExe formatter;
+          in
+            drv + "\n" + bin + "\n"
+        else
+          ""
+    ]]
+
+    cp = vim.system({
+        "nix",
+        "--extra-experimental-features",
+        "nix-command flakes",
+        "eval",
+        ".#formatter",
+        "--raw",
+        "--apply",
+        eval_nix_formatter,
+    }, { cwd = params.root }):wait()
+    if cp.code ~= 0 then
+        -- Dirty hack to check if the flake actually defines a formatter.
+        --
+        -- I cannot for the *life* of me figure out a less hacky way of
+        -- checking if a flake defines a formatter. Things I've tried:
+        --
+        --   - `nix eval . --apply '...'`: This doesn't not give me the flake
+        --                                 itself, it gives me the default package.
+        --   - `builtins.getFlake`: Every incantation I've tried requires
+        --                          `--impure`, which has the performance downside described above.
+        --   - `nix flake show --json .`: This works, but it can be quite slow:
+        --                                we end up evaluating all outputs, which can take a while for
+        --                                `nixosConfigurations`.
+        if cp.stderr:find("error: flake .+ does not provide attribute .+ or 'formatter'") then
+            log:warn("this flake does not define a `nix fmt` entrypoint")
+        else
+            log:error(string.format("unable discover 'nix fmt' command. stderr: %s", cp.stderr))
+        end
```

Comment:
I don't know of any better way to check if a flake exposes an attr. but can't we just forward the error as warning in any case this fails and hope that nix provides a good enough warning so we don't need to handle it explicitly?

Reply:
probably warn everytime?

Reply:
That's fair! I've updated this to warn instead.